[openssl] build failure

Package: openssl[core]:x64-windows -> 3.1.0#1

**Host Environment**

- Host: x64-windows
- Compiler: MSVC 19.35.32216.1
-    vcpkg-tool version: 2023-03-29-664f8bb619b752430368d0f30a8289b761f5caba
    vcpkg-scripts version: b619a233f 2023-04-03 (6 hours ago)

**To Reproduce**

`./vcpkg.exe install openssl:x64-windows`

**Failure logs**

```
-- Using cached openssl-openssl-openssl-3.1.0.tar.gz.
-- Cleaning sources at C:/Code/repo/github/microsoft/vcpkg/buildtrees/openssl/src/nssl-3.1.0-1ebd9e680e.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/Code/repo/github/microsoft/vcpkg/downloads/openssl-openssl-openssl-3.1.0.tar.gz
-- Applying patch disable-apps.patch
-- Applying patch disable-install-docs.patch
-- Applying patch script-prefix.patch
-- Applying patch windows/install-layout.patch
-- Applying patch windows/install-pdbs.patch
-- Applying patch unix/android-cc.patch
-- Applying patch unix/move-openssldir.patch
-- Applying patch unix/no-empty-dirs.patch
-- Applying patch unix/no-static-libs-for-shared.patch
-- Using source at C:/Code/repo/github/microsoft/vcpkg/buildtrees/openssl/src/nssl-3.1.0-1ebd9e680e.clean
-- Found external ninja('1.11.0').
-- Getting CMake variables for x64-windows
-- Getting CMake variables for x64-windows
-- Downloading https://download.qt.io/official_releases/jom/jom_1_1_3.zip;https://mirrors.ocf.berkeley.edu/qt/official_releases/jom/jom_1_1_3.zip -> jom_1_1_3.zip...
-- Prerunning x64-windows-dbg
-- Building x64-windows-dbg
-- Restarting build without parallelism
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.35.32215/bin/Hostx64/x64/nmake.exe" /NOLOGO /G /U /F makefile install_dev install_modules INSTALL_PDBS=ON
    Working Directory: C:/Code/repo/github/microsoft/vcpkg/buildtrees/openssl/x64-windows-dbg
    See logs for more information:
      C:\Code\repo\github\microsoft\vcpkg\buildtrees\openssl\install-x64-windows-dbg-nmake-out.log
      C:\Code\repo\github\microsoft\vcpkg\buildtrees\openssl\install-x64-windows-dbg-nmake-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_nmake.cmake:156 (vcpkg_execute_build_process)
  ports/openssl/windows/portfile.cmake:50 (vcpkg_build_nmake)
  ports/openssl/portfile.cmake:62 (include)
  scripts/ports.cmake:147 (include)
```

[install-x64-windows-dbg-nmake-err.log](https://github.com/microsoft/vcpkg/files/11142964/install-x64-windows-dbg-nmake-err.log)
[install-x64-windows-dbg-nmake-out.log](https://github.com/microsoft/vcpkg/files/11142966/install-x64-windows-dbg-nmake-out.log)

**Additional context**

- Operating system: Win 11, Version 22H2 (OS Build 22621.1485)
- Visual Studio 2022: 17.5.3 (Did clean installation)
- Windows 11 SDK (10.0.22000.0). Tried also (10.0.22621.0), and the same thing happened.
- Tried to apply almost every suggested solution for  `openssl` issues on `github`, which I find relative. Nothing helped. Still getting an error as listed above.
- Uninstalled the global `CMake` version by allowing `vcpkg` to download and use its CMake version. Ran the build on a new windows terminal. I tried with default Powersell, the latest Powersell, and from cmd.
- I have issues only with building openssl. I can successfully build, for example, `gtest`, `spdlog`, `boost`.
- Every time before execution of `vcpkg install ...`  command, I have deleted everything generated previously.

--- a/ports/openssl/portfile.cmake
+++ b/ports/openssl/portfile.cmake
@@ -26,6 +26,7 @@ vcpkg_from_github(
         script-prefix.patch
         windows/install-layout.patch
         windows/install-pdbs.patch
+        windows/perlasm-scheme.patch
         unix/android-cc.patch
         unix/move-openssldir.patch
         unix/no-empty-dirs.patch

--- a/ports/openssl/vcpkg.json
+++ b/ports/openssl/vcpkg.json
@@ -1,6 +1,7 @@
 {
   "name": "openssl",
   "version": "3.1.2",
+  "port-version": 1,
   "description": "OpenSSL is an open source project that provides a robust, commercial-grade, and full-featured toolkit for the Transport Layer Security (TLS) and Secure Sockets Layer (SSL) protocols. It is also a general-purpose cryptography library.",
   "homepage": "https://www.openssl.org",
   "license": "Apache-2.0",

--- a/ports/openssl/windows/perlasm-scheme.patch
+++ b/ports/openssl/windows/perlasm-scheme.patch
@@ -1,0 +1,39 @@
+diff --git a/Configurations/10-main.conf b/Configurations/10-main.conf
+index e8084ab..263ecd0 100644
+--- a/Configurations/10-main.conf
++++ b/Configurations/10-main.conf
+@@ -9,19 +9,22 @@ sub vc_win64a_info {
+             $vc_win64a_info = { AS        => "nasm",
+                                 ASFLAGS   => "-g",
+                                 asflags   => "-Ox -f win64 -DNEAR",
+-                                asoutflag => "-o " };
++                                asoutflag => "-o ",
++                                perlasm_scheme => "nasm" };
+         } elsif ($disabled{asm}) {
+             # assembler is still used to compile uplink shim
+             $vc_win64a_info = { AS        => "ml64",
+                                 ASFLAGS   => "/nologo /Zi",
+                                 asflags   => "/c /Cp /Cx",
+-                                asoutflag => "/Fo" };
++                                asoutflag => "/Fo",
++                                perlasm_scheme => "masm" };
+         } else {
+             $die->("NASM not found - make sure it's installed and available on %PATH%\n");
+             $vc_win64a_info = { AS        => "{unknown}",
+                                 ASFLAGS   => "",
+                                 asflags   => "",
+-                                asoutflag => "" };
++                                asoutflag => "",
++                                perlasm_scheme => "auto" };
+         }
+     }
+     return $vc_win64a_info;
+@@ -1493,7 +1496,7 @@ my %targets = (
+         sys_id           => "WIN64A",
+         uplink_arch      => 'x86_64',
+         asm_arch         => 'x86_64',
+-        perlasm_scheme   => "auto",
++        perlasm_scheme   => sub { vc_win64a_info()->{perlasm_scheme} },
+         multilib         => "-x64",
+     },
+     "VC-WIN32" => {

--- a/versions/baseline.json
+++ b/versions/baseline.json
@@ -6082,7 +6082,7 @@
     },
     "openssl": {
       "baseline": "3.1.2",
-      "port-version": 0
+      "port-version": 1
     },
     "openssl-unix": {
       "baseline": "deprecated",

--- a/versions/o-/openssl.json
+++ b/versions/o-/openssl.json
@@ -1,6 +1,11 @@
 {
   "versions": [
     {
+      "git-tree": "5a99562f8ac47a68a58efba5f0d40259fedeecc5",
+      "version": "3.1.2",
+      "port-version": 1
+    },
+    {
       "git-tree": "db4486fff3146088dd150f067ebf940c649ffe42",
       "version": "3.1.2",
       "port-version": 0

